File: 98-validate-binary-search-tree/98-validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool Traverse(TreeNode* root, long low, long high){
        
        if(root != NULL){
        
            if(root->val >= high || root->val <= low )
                return false;

            return Traverse(root->left, low, root->val) && Traverse(root->right, root->val, high);
        }
        
        return true;
    }
    
    bool isValidBST(TreeNode* root) {
        
        if(root == NULL)
            return true;
        
        return Traverse(root, LONG_MIN, LONG_MAX);
        
    }
};

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def traverse(self, root, high, low):
        
        if root == None:
            return True
        
        if root.val >= high or root.val <= low:
            return False
        
        return self.traverse(root.right, high, root.val) and self.traverse(root.left, root.val, low)
        
    
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        
        high = 2**31 + 1
        low = -2**31 - 1
        
        return self.traverse(root.left, root.val, low) and self.traverse(root.right, high, root.val)